api/service/projects_service.go

```diff
@@ -250,3 +283,116 @@ func (service *projectsService) filterAuthorizedProjects(ctx context.Context, pr
+		service.updateProjectConfig.ResponseTemplate == "" {
+		return project, nil, nil
+	}
+
+	payload, err := generateRequestPayload(project, service.updateProjectConfig.PayloadTemplate)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error generating request payload: %w", err)
+	}
+
+	resp, err := sendUpdateRequest(service.updateProjectConfig.Endpoint, payload)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error sending update request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	response, err := processResponseTemplate(resp, service.updateProjectConfig.ResponseTemplate)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error processing response template: %w", err)
+	}
+
+	return project, response, nil
+}
+
+func generateRequestPayload(project *models.Project, templateString string) (map[string]interface{}, error) {
+	tmpl, err := template.New("requestPayload").Parse(templateString)
+	if err != nil {
+		return nil, err
+	}
+	var payload bytes.Buffer
+	if err := tmpl.Execute(&payload, project); err != nil {
+		return nil, err
+	}
+	var result map[string]interface{}
+	if err := json.Unmarshal(payload.Bytes(), &result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func sendUpdateRequest(url string, payload map[string]interface{}) (*http.Response, error) {
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payloadBytes))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+func processResponseTemplate(response *http.Response, templateString string) (map[string]interface{}, error) {
+	var data map[string]interface{}
+	if err := json.NewDecoder(response.Body).Decode(&data); err != nil {
+		return nil, err
+	}
+
+	tmpl, err := template.New("responsePayload").Parse(templateString)
+	if err != nil {
+		return nil, err
+	}
+
+	var responseText bytes.Buffer
+	if err := tmpl.Execute(&responseText, data); err != nil {
+		return nil, err
+	}
+
+	var result map[string]interface{}
+	if err := json.Unmarshal(responseText.Bytes(), &result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// isLabelBlacklisted check if any key in labels is blacklisted
+func isLabelBlacklisted(existingLabels, newLabels []models.Label, blacklist []string) bool {
+	for _, newLabel := range newLabels {
+		if contain(newLabel.Key, blacklist) {
+			for _, existingLabel := range existingLabels {
+				if existingLabel.Key == newLabel.Key && existingLabel.Value != newLabel.Value {
+					return true
+				}
+			}
+		} else {
+			return false
+		}
+	}
+	return false
+}
+
+// contain function to check if a string is in a slice
+func contain(item string, slice []string) bool {
+	for _, label := range slice {
+		if label == item {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
In order not to do O(n) list lookups (where n is the number of elements in the blacklist) each time this function is called, can we instead do O(1) lookups by making `blacklist` a set/map instead? This can also involve refactoring `service.updateProjectConfig.LabelsBlacklist` a little to make sure it's stored as a set/map and not a list.